src/net.rs

```diff
@@ -0,0 +1,331 @@
+pub trait HostResponse {
+    /// Returns the raw byte stream for building the payload of the response.
+    ///
+    /// This function should not be called after calling `finish()`.
+    fn sink(&mut self) -> Result<&mut dyn Write, Error>;
+
+    /// Indicates that all payload data has been written.
+    ///
+    /// Callers should remember to call this function; failing to do so may
+    /// result in a response not being sent properly.
+    fn finish(&mut self) -> Result<(), Error>;
+}
+
+/// A simple in-memory [`HostPort`].
+///
+/// This type is both useful for testing, and as a demonstration of how to use
+/// a [`HostPort`] and its associated traits.
+///
+/// [`HostPort`]: trait.HostPort.html
+///
+/// # Example
+/// ```
+/// # use manticore::io::*;
+/// # use manticore::mem::*;
+/// # use manticore::net::*;
+/// # use manticore::protocol::*;
+/// # use manticore::protocol::firmware_version::*;
+/// # use manticore::protocol::wire::*;
+/// // Build the InMemHost.
+/// let mut buf = [0; 64];
+/// let mut host = InMemHost::new(&mut buf);
+///
+/// // Prepare a request to push into the host.
+/// let header = Header {
+///     command: CommandType::FirmwareVersion,
+///     is_request: true,
+/// };
+/// let req = [0];
+/// host.request(header, &req);
+///
+/// // Set up an arena for parsing.
+/// let mut arena = [0; 64];
+/// let mut arena = BumpArena::new(&mut arena);
+///
+/// // Prepare to recieve a message.
+/// let mut host_req = host.receive()?;
+/// let header = host_req.header()?;
+/// assert_eq!(header.command, CommandType::FirmwareVersion);
+/// assert!(header.is_request);
+///
+/// // Parse and process the message.
+/// let req = FirmwareVersionRequest::from_wire(
+///     host_req.payload()?, &arena).unwrap();
+/// assert_eq!(req.index, 0);
+///
+/// // Prepare to reply to the message.
+/// let mut host_resp = host_req.reply(Header {
+///     command: CommandType::FirmwareVersion,
+///     is_request: false,
+/// })?;
+///
+/// // Build and write a reply.
+/// let resp = FirmwareVersionResponse {
+///     version: &[0xba; 32],
+/// };
+/// resp.to_wire(host_resp.sink()?).unwrap();
+/// host_resp.finish()?;
+///
+/// // Check that we got the right data back.
+/// let (header, resp_bytes) = host.response().unwrap();
+/// assert_eq!(header.command, CommandType::FirmwareVersion);
+/// assert!(!header.is_request);
+///
+/// // Now, parse the response.
+/// arena.reset();
+/// let resp = FirmwareVersionResponse::from_wire(resp_bytes, &arena).unwrap();
+/// assert_eq!(resp.version, &[0xba; 32]);
+/// # Ok::<(), manticore::net::Error>(())
+/// ```
+pub struct InMemHost<'buf>(InMemInner<'buf>);
+
+/// The actual guts of an `InMemHost`. This struct is used to implement the two
+/// "connection state" traits used by `HostPort`.
+struct InMemInner<'buf> {
+    rx_header: Option<Header>,
+    rx: &'buf [u8],
+    tx_header: Option<Header>,
+    tx: Cursor<'buf>,
+    finished: bool,
+}
```

Comment:
I don't understand why this nesting is necessary from the provided examples. The tests compile and pass if `InMemInner` is removed and instead these fields are members of `InMemHost`, and `HostRquest` and `HostReceive` are implemented directly on `InMemHost`.

Reply:
I added a comment that explains this indirection.